sql: panic with nullable repeated projections in full outer joins

This query causes the database to crash:

`SELECT *, pg_typeof(column1) FROM (SELECT * FROM (VALUES (NULL, NULL)) NATURAL FULL OUTER JOIN (SELECT * FROM (VALUES (1, 1))));`

There is no error message. The server process ends, and the client says "connection lost".

Thanks for the report! This usually means the server has panicked - you should be able to find that message in the server logs.

I went ahead and reproduced it locally. Here's the panic:

```
panic: invalid datum type given: int, expected unknown

goroutine 238 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0xd, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6097940, 0xc425bca080, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:140 +0x23f
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcOutputHelper).ProcessRow(0xc425ef8400, 0x6077aa0, 0xc425ede740, 0xc425cccc30, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:330 +0x6c6
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*noopProcessor).Next(0xc425ef8400, 0x0, 0x0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:543 +0xca
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x6077aa0, 0xc425ede740, 0x6078860, 0xc425ef8400, 0x605e0e0, 0xc425ebc900)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:146 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*noopProcessor).Run(0xc425ef8400, 0xc42001ddf8)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:499 +0x6b
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:466 +0x38b
```

--- a/pkg/sql/join.go
+++ b/pkg/sql/join.go
@@ -292,7 +292,12 @@ func (p *planner) makeJoin(
 				return planDataSource{}, err
 			}
 		}
-		r.addRenderColumn(expr, symbolicExprStr(expr), left.info.SourceColumns[leftCol])
+		// Issue #23609: the type of the left side might be NULL; so use the type of
+		// the IFNULL expression instead of the type of the left source column.
+		r.addRenderColumn(expr, symbolicExprStr(expr), sqlbase.ResultColumn{
+			Name: left.info.SourceColumns[leftCol].Name,
+			Typ:  expr.ResolvedType(),
+		})
 	}
 	for i, c := range left.info.SourceColumns {
 		if remapped[i] != -1 {

--- a/pkg/sql/join_predicate.go
+++ b/pkg/sql/join_predicate.go
@@ -113,7 +113,8 @@ func (p *joinPredicate) tryAddEqualityFilter(
 	}
 
 	if !lhs.ResolvedType().Equivalent(rhs.ResolvedType()) {
-		// We can't have equality columns of different types (#22519).
+		// Issue #22519: we can't have two equality columns of mismatched types
+		// because the hash-joiner assumes the encodings are the same.
 		return false
 	}
 

--- a/pkg/sql/logictest/testdata/logic_test/join
+++ b/pkg/sql/logictest/testdata/logic_test/join
@@ -2345,3 +2345,34 @@ NULL  44    51
 NULL  NULL  52
 NULL  42    53
 NULL  45    45
+
+# Regression test for #23609: make sure that the type of the merged column
+# is int (not unknown).
+query TT colnames
+SELECT "Type", "Columns" FROM [EXPLAIN (TYPES)
+  SELECT column1, column1+1
+  FROM
+    (SELECT * FROM
+      (VALUES (NULL, NULL)) AS t
+        NATURAL FULL OUTER JOIN
+      (VALUES (1, 1)) AS u)]
+WHERE "Type" != ''
+----
+Type    Columns
+render  (column1 int, "column1 + 1" int)
+render  (column1 int, column2[omitted] int)
+render  (column1 int, column2[omitted] int, column1[hidden,omitted] unknown, column2[hidden,omitted] unknown, column1[hidden,omitted] int, column2[hidden,omitted] int)
+join    (column1 unknown, column2[omitted] unknown, column1 int, column2[omitted] int)
+values  (column1 unknown, column2 unknown)
+values  (column1 int, column2 int)
+
+query II rowsort
+SELECT column1, column1+1
+FROM
+  (SELECT * FROM
+    (VALUES (NULL, NULL)) AS t
+      NATURAL FULL OUTER JOIN
+    (VALUES (1, 1)) AS u)
+----
+1     2
+NULL  NULL

